packages/react-theme-provider/src/ThemeProvider.tsx

```diff
@@ -0,0 +1,55 @@
+import * as React from 'react';
+import cx from 'classnames';
+import { useStylesheet } from '@fluentui/react-stylesheets';
+import { tokensToStyleObject } from './tokensToStyleObject';
+import { ThemeContext } from './ThemeContext';
+import { Theme, ThemePrepared } from './types';
+import { mergeThemes } from './mergeThemes';
+import { useTheme } from './useTheme';
+import * as classes from './ThemeProvider.scss';
+
+/**
+ * Props for the ThemeProvider component.
+ */
+export interface ThemeProviderProps extends React.HTMLAttributes<HTMLDivElement> {
+  /**
+   * Defines the theme provided by the user.
+   */
+  theme?: Theme;
+}
+
+/**
+ * ThemeProvider, used for providing css variables and registering stylesheets.
+ */
+export const ThemeProvider = React.forwardRef<HTMLDivElement, ThemeProviderProps>(
+  (
+    { theme, style, className, ...rest }: React.PropsWithChildren<ThemeProviderProps>,
+    ref: React.Ref<HTMLDivElement>,
+  ) => {
+    // Pull contextual theme.
+    const parentTheme = useTheme();
+
+    // Merge the theme only when parent theme or props theme mutates.
+    const fullTheme = React.useMemo<ThemePrepared>(() => mergeThemes(parentTheme, theme), [parentTheme, theme]);
```

Comment:
As `theme` is a prop it can be inline object: `<ThemeProvider theme={{ foo: 'bar' }} />`. It means that all consumers will receive updates on each render as `fullTheme` is passed to React context.